File: app/src/main/java/com/makingiants/today/todayhistory/utils/log/CrashlyticsTree.java
package com.makingiants.today.todayhistory.utils.log;

import android.content.Context;
import android.util.Log;
import com.crashlytics.android.Crashlytics;
import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.core.CrashlyticsCore;
import com.makingiants.today.todayhistory.BuildConfig;
import io.fabric.sdk.android.Fabric;
import java.util.Map;
import timber.log.Timber;

public class CrashlyticsTree extends Timber.Tree {
  public CrashlyticsTree(Context context, Map<String, String> properties) {
    Fabric.with(context, new Crashlytics(), new Answers());

    CrashlyticsCore core = Crashlytics.getInstance().core;

    for (Map.Entry<String, String> entry : properties.entrySet()) {
      core.setString(entry.getKey(), entry.getValue());
    }
  }

  @Override
  protected void log(int priority, String tag, String message, Throwable t) {
    if (priority == Log.VERBOSE && !"release".equals(BuildConfig.BUILD_TYPE)) {
      //Dont log verbose messages and avoid not release builds to send crash to crashlytics
      return;
    }

    CrashlyticsCore core = Crashlytics.getInstance().core;
    core.log(Log.ERROR, tag, message);

    if (t != null && priority == Log.ERROR) {
      core.logException(t);
    }
  }
}

File: app/src/main/kotlin/com/makingiants/todayhistory/utils/log/CrashlyticsTree.kt
package com.makingiants.todayhistory.utils.log

import android.content.Context
import android.util.Log
import com.crashlytics.android.Crashlytics
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.core.CrashlyticsCore
import com.makingiants.todayhistory.BuildConfig
import io.fabric.sdk.android.Fabric
import timber.log.Timber

class CrashlyticsTree(context: Context, properties: Map<String, String>) : Timber.Tree() {
    init {
        Fabric.with(context, Crashlytics(), Answers())

        val core = Crashlytics.getInstance().core

        for (entry in properties.entries) {
            core.setString(entry.key, entry.value)
        }
    }

    override fun log(priority: Int, tag: String, message: String, t: Throwable?) {
        if (priority == Log.VERBOSE && "release" != BuildConfig.BUILD_TYPE) {
            //Dont log verbose messages and avoid not release builds to send crash to crashlytics
            return
        }

        val core = Crashlytics.getInstance().core
        core.log(Log.ERROR, tag, message)

        if (t != null && priority == Log.ERROR) {
            core.logException(t)
        }
    }
}
